### PR TITLE
Allow call caching based on singularity paths [BA-6440]

### PR DESCRIPTION
**Pull request by cjllanwarne**
This wasn't as simple as just a one line "add singularity to the list" of cacheable attributes, because `singularity` is a user-generated attribute specified in the Config backend's configuration. So the change ended up being:

* Allow additional config allowing users to specify which runtime attributes are call-cacheable
* Wiring that through the runtime attribute validators to include the correct call-cacheable-ness
* Unit and centaur tests.

### CODE REVIEW

**Comment by cjllanwarne** [comment]
cc @rhpvorderman @gbggrant who might be interested in the singularity aspects of this change to config backend.

File: src/ci/resources/local_application.conf [hunk 1]
```diff
@@ -28,5 +28,23 @@ backend {
...
+      config {
+        include "local_provider_config.inc.conf"
+        runtime-attributes = """
+          Int cacheworthy_attribute
+          Int uncacheworthy_attribute_1
+          Int uncacheworthy_attribute_2
+        """
+        runtime-attributes-for-caching = {
+          cacheworthy_attribute: true
+          cacheworthy_attribute_1: true
```
Comment (grsterin) [inline src/ci/resources/local_application.conf:42]: Was it meant to be `uncacheworthy_attribute_1: false`?

File: centaur/src/main/resources/standardTestCases/custom_cacheworthy_attributes/custom_cacheworthy_attributes.wdl [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+        # Config for this backend is defined in src/ci/resources/local_application.conf
+        backend: "LocalCacheableRuntimeAttribute"
+    }
+    command {
+      echo foo > foo
+    }
+    output {
+        String foo = read_string("foo")
+    }
+}
```
Comment (grsterin) [inline centaur/src/main/resources/standardTestCases/custom_cacheworthy_attributes/custom_cacheworthy_attributes.wdl:60]: \n
```suggestion
}

```

File: cromwell.example.backends/singularity.conf [hunk 3]
```diff
@@ -14,17 +14,18 @@ backend {
...
-            submit-docker = """
+        submit-docker = """
               singularity exec --bind ${cwd}:${docker_cwd} docker://${docker} ${job_shell} ${script}
             """
-        }
+      }
     }
-}
+  }
+}
```
Comment (grsterin) [inline cromwell.example.backends/singularity.conf:31]: \n
```suggestion
}

```

File: src/ci/resources/local_application.conf (hunk 1 above)
Comment (cjllanwarne) [reply src/ci/resources/local_application.conf:42]: oops, yes! Well spotted!

File: docs/tutorials/HPCIntro.md [hunk 4]
```diff
@@ -132,6 +132,31 @@ backend.providers.SGE.config {
...
+  # ...
+  """
+  runtime-attributes-for-caching {
+    sge_queue: false
+    singularity_image: true
+  }
+}
+```
+
+Note: Only *custom* attributes can be altered like this. Memory, CPU and docker will always have their default cache-consideration behavior.
```
Comment (kshakir) [inline docs/tutorials/HPCIntro.md:158]: I'd also make a note that unlike `runtime { docker: ... }`, this PR _only_ checks that the `runtime { singularity: ... }` string is the same cached string.

The `docker` runtime attributes go through an extra effort, even on the local backend, dereferencing the image-name-string converting it to a hash. This happens using the on-machine docker daemon. This type of lookup will NOT happen for `singularity` image names, meaning if you tag a new image as "my-latest-version" you will still end up call-caching even though a new image available.

https://github.com/broadinstitute/cromwell/blob/51/dockerHashing/src/test/scala/cromwell/docker/local/DockerCliSpec.scala

One could build that extra-hop plugin interface... but I really think a warning is enough for now / iterative development.

**Comment by rhpvorderman** [comment]
@cjllanwarne Thanks for notifying! We always use the `docker` attribute in BioWDL, so BioWDL pipelines can run without any extra configuration. On our cluster we have configured this so that we run the docker images using singularity. 

I think it is a good thing that custom runtime attributes can be cached now. We recently added a `time_minutes` attribute to our pipelines in order to work better with SLURM. I hope this code ties in nicely when we switch to `hints` in WDL 2.0.

File: supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/DeclarationValidation.scala [hunk 5]
```diff
@@ -84,10 +91,15 @@ class DeclarationValidation(declaration: Declaration, instanceValidation: Runtim
...
-      default(instanceValidation, declaration.expression.get)
-    else instanceValidation
-    if (declaration.womType.isInstanceOf[WomOptionalType]) validationDefault.optional else validationDefault
+    import scala.language.existentials
+    val validationWithDefault = if (declaration.expression.isDefined) default(instanceValidation, declaration.expression.get) else instanceValidation
+    val validationWithDefaultAndOptionality = if (declaration.womType.isInstanceOf[WomOptionalType]) validationWithDefault.optional else validationWithDefault
+    val validationWithDefaultAndOptionalityAndCallCaching = usedInCallCachingOverride match {
+      case Some(usedInCallCachingValue) => RuntimeAttributesValidation.withUsedInCallCaching(validationWithDefaultAndOptionality, usedInCallCachingValue)
+      case None => validationWithDefaultAndOptionality
+    }
```
[the remark this answers is not in the document]
Comment (cjllanwarne) [reply supportedBackends/sfs/src/main/scala/cromwell/backend/impl/sfs/config/DeclarationValidation.scala:100]: I'll try!

**Pull request merged**